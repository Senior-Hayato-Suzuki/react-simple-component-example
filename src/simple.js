import React from "react";
import cn from "classnames";
import {css} from "glamor";
import omit from "lodash/fp/omit";
import pick from "lodash/fp/pick";
import pickBy from "lodash/fp/pickBy";
import deepmerge from "deepmerge";

const pickTruthty = pickBy(Boolean);

function getDisplayName(Component) {
    if (typeof Component === "string") {
        return Component;
    }

    return Component.displayName || Component.name || "Anonymous";
}

var compnentCounter = 0;

function simple(Component, styles, alts={}, options={}) {
    const num = compnentCounter++;
    const makeDebugClass = className => className
            ? className + "___" + num
            : null;

    const level = options._level || 0;

    if (Component._styleWrapped) {
        return simple(
            Component._styleWrapped.Component,
            deepmerge(Component._styleWrapped.styles, styles),
            deepmerge(Component._styleWrapped.alts, alts),
            {
                _level: level + 1,
            }
        );
    }

    const rules = {__base: css(styles)};

    const altProps = Object.keys(alts);


    for (let key in alts) {
        rules[key] = css(deepmerge(styles, alts[key]));
    }

    function StyleWrapped({className, ...otherProps}) {

        const passProps = omit(altProps, otherProps);
        const alt = Object.keys(pickTruthty(pick(altProps, otherProps)));


        if (alt.length > 1) {
            throw new Error("Too many alt props: " + alt.join(", "));
        }

        const rule = rules[alt[0] || "__base"];


        const props = {
            ...passProps,
            className: cn(makeDebugClass(StyleWrapped.displayName), makeDebugClass(alt[0]), className, String(rule)),
        };

        return <Component {...props} />;
    }

    StyleWrapped._styleWrapped = {
        Component,
        styles,
        alts,
    };

    StyleWrapped.displayName = "simple(" + getDisplayName(Component) + ")";

    StyleWrapped.create = el => simple(el, styles, alts);


    return StyleWrapped;
}


const viewStyles = {
    boxSizing: "border-box",
    position: "relative",
    display: "flex",
    flexDirection: "column",
    alignItems: "stretch",
    flexShrink: 0,
    alignContent: "flex-start",
    border: 0,
    margin: 0,
    padding: 0,
    minWidth: 0,
};

export const View = simple("div", viewStyles);
View.displayName = "View";


export {css};
export default simple;
