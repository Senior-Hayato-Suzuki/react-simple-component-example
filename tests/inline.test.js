import React from "react";
import renderer from "react-test-renderer";

import simple from "../src/inline";

describe("inline backend", () => {
    test("renders style attribute and debug class name", () => {
        const Button = simple("button", {
            padding: 10,
        });

        const tree = renderer.create(<Button />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("passes custom classes ", () => {
        const Button = simple("button", {
            padding: 10,
        });

        const tree = renderer
            .create(<Button className="custom-class-name" />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("passes custom class names", () => {
        const Button = simple("button", {
            padding: 10,
        });

        const tree = renderer
            .create(<Button className="custom-class-name" />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("renders the alt styles and debug class", () => {
        const Button = simple(
            "button",
            {
                padding: 10,
            },
            {
                red: {
                    backgroundColor: "red",
                },
            },
        );
        Button; // eslint-disable-line

        const tree = renderer.create(<Button red />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("renders the combined styles for extended components", () => {
        const Button = simple("button", {
            padding: 10,
        });
        const RoundButton = simple(Button, {
            borderRadius: 10,
        });

        const tree = renderer.create(<RoundButton />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
