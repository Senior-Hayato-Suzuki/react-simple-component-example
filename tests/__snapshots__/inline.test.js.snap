exports[`inline backend passes custom class names 1`] = `
<button
  className="Button___1 custom-class-name"
  style={
    Object {
      "padding": 10,
    }
  } />
`;

exports[`inline backend renders style attribute and debug class name 1`] = `
<button
  className="Button___0"
  style={
    Object {
      "padding": 10,
    }
  } />
`;

exports[`inline backend renders the alt styles and debug class 1`] = `
<button
  className="Button___2 red___2"
  style={
    Object {
      "backgroundColor": "red",
      "padding": 10,
    }
  } />
`;

exports[`inline backend renders the combined styles for extended components 1`] = `
<button
  className="RoundButton___3"
  style={
    Object {
      "borderRadius": 10,
      "padding": 10,
    }
  } />
`;
